File: ocaml/infer.ml
open Ast
open Types

type env = (string * typ) list

let lookup x env =
  try List.assoc x env with Not_found -> failwith ("unbound variable " ^ x)

let fresh_float () = TFloat (fresh_mode_meta ())

let ensure_float expected =
  match zonk expected with
  | TFloat m -> TFloat m
  | TMeta m ->
      let mvar = fresh_mode_meta () in
      set_type m (TFloat mvar); TFloat mvar
  | _ -> failwith "expected float type"

let rec infer (env : env) (exp : Ast.expr) (expected : typ) : typed_expr =
  match exp with
  | Var x ->
      let ty_var = lookup x env in
      assert_subtype ty_var expected;
      { expr = EVar x; typ = expected }
  | Lam (x, body) ->
      let dom = TMeta (fresh_meta ()) in
      let cod = TMeta (fresh_meta ()) in
      let body_t = infer ((x, dom) :: env) body cod in
      let lam_ty = TArrow (dom, cod) in
      assert_subtype lam_ty expected;
      { expr = ELam (x, body_t); typ = lam_ty }
  | Rec (f, x, body) ->
      let dom = TMeta (fresh_meta ()) in
      let cod = TMeta (fresh_meta ()) in
      let fn_ty = TArrow (dom, cod) in
      let env' = (f, fn_ty) :: (x, dom) :: env in
      let body_t = infer env' body cod in
      assert_subtype body_t.typ cod;
      assert_subtype fn_ty expected;
      { expr = ERec (f, x, body_t); typ = fn_ty }
  | App (e1, e2) ->
      let arg_ty = TMeta (fresh_meta ()) in
      let res_ty = TMeta (fresh_meta ()) in
      let fn_ty = TArrow (arg_ty, res_ty) in
      let e1_t = infer env e1 fn_ty in
      let e2_t = infer env e2 arg_ty in
      assert_subtype res_ty expected;
      { expr = EApp (e1_t, e2_t); typ = res_ty }
  | Unit ->
      assert_subtype TUnit expected;
      { expr = EUnit; typ = TUnit }
  | Nil ->
      let elem = TMeta (fresh_meta ()) in
      let list_ty = TList elem in
      assert_subtype list_ty expected;
      { expr = ENil; typ = list_ty }
  | Cons (hd, tl) ->
      let elem = TMeta (fresh_meta ()) in
      let list_ty = TList elem in
      let hd_t = infer env hd elem in
      let tl_t = infer env tl list_ty in
      assert_subtype list_ty expected;
      { expr = ECons (hd_t, tl_t); typ = list_ty }
  | Pair (e1, e2) ->
      let t1 = TMeta (fresh_meta ()) in
      let t2 = TMeta (fresh_meta ()) in
      let e1_t = infer env e1 t1 in
      let e2_t = infer env e2 t2 in
      let pair_ty = TPair (e1_t.typ, e2_t.typ) in
      assert_subtype pair_ty expected;
      { expr = EPair (e1_t, e2_t); typ = pair_ty }
  | Fst e ->
      let a_ty = TMeta (fresh_meta ()) in
      let b_ty = TMeta (fresh_meta ()) in
      let pair_ty = TPair (a_ty, b_ty) in
      let e_t = infer env e pair_ty in
      assert_subtype a_ty expected;
      { expr = EFst e_t; typ = a_ty }
  | Snd e ->
      let a_ty = TMeta (fresh_meta ()) in
      let b_ty = TMeta (fresh_meta ()) in
      let pair_ty = TPair (a_ty, b_ty) in
      let e_t = infer env e pair_ty in
      assert_subtype b_ty expected;
      { expr = ESnd e_t; typ = b_ty }
  | Inl e ->
      let left_ty = TMeta (fresh_meta ()) in
      let right_ty = TMeta (fresh_meta ()) in
      let e_t = infer env e left_ty in
      let sum_ty = TSum (e_t.typ, right_ty) in
      assert_subtype sum_ty expected;
      { expr = EInl e_t; typ = sum_ty }
  | Inr e ->
      let left_ty = TMeta (fresh_meta ()) in
      let right_ty = TMeta (fresh_meta ()) in
      let e_t = infer env e right_ty in
      let sum_ty = TSum (left_ty, e_t.typ) in
      assert_subtype sum_ty expected;
      { expr = EInr e_t; typ = sum_ty }
  | Case (e, (x, e1), (y, e2)) ->
      let l_ty = TMeta (fresh_meta ()) in
      let r_ty = TMeta (fresh_meta ()) in
      let scrut_ty = TSum (l_ty, r_ty) in
      let scrut = infer env e scrut_ty in
      let e1_t = infer ((x, l_ty) :: env) e1 expected in
      let e2_t = infer ((y, r_ty) :: env) e2 expected in
      { expr = EMatch (scrut, (x, e1_t), (y, e2_t)); typ = expected }
  | MatchList (e, nil_br, (x, xs, cons_br)) ->
      let elem_ty = TMeta (fresh_meta ()) in
      let list_ty = TList elem_ty in
      let scrut = infer env e list_ty in
      let nil_t = infer env nil_br expected in
      let cons_env = (x, elem_ty) :: (xs, list_ty) :: env in
      let cons_t = infer cons_env cons_br expected in
      { expr = EMatchList (scrut, nil_t, (x, xs, cons_t)); typ = expected }
  | Bool b ->
      assert_subtype TBool expected;
      { expr = EBool b; typ = TBool }
  | If (c, t, f) ->
      let c_t = infer env c TBool in
      let t_t = infer env t expected in
      let f_t = infer env f expected in
      { expr = EIf (c_t, t_t, f_t); typ = expected }
  | Let (x, e1, e2) ->
      let ty1 = TMeta (fresh_meta ()) in
      let e1_t = infer env e1 ty1 in
      let e2_t = infer ((x, e1_t.typ) :: env) e2 expected in
      { expr = ELet (x, e1_t, e2_t); typ = e2_t.typ }
  | Const f ->
      let ty = ensure_float expected in
      { expr = EConst f; typ = ty }
  | Neg e ->
      let ty = ensure_float expected in
      let e_t = infer env e ty in
      { expr = ENeg e_t; typ = ty }
  | Add (a, b) ->
      let ty = ensure_float expected in
      let a_t = infer env a ty in
      let b_t = infer env b ty in
      { expr = EAdd (a_t, b_t); typ = ty }
  | Sub (a, b) ->
      let ty = ensure_float expected in
      let a_t = infer env a ty in
      let b_t = infer env b ty in
      { expr = ESub (a_t, b_t); typ = ty }
  | Mul (a, b) ->
        let is_scaling = match (a,b) with (Const _, _) | (_, Const _) -> true | _ -> false in
        let a_ty, b_ty = 
            if is_scaling 
            then 
                expected, expected
            else
                let g_mode = fresh_mode_meta () in
                set_mode g_mode G;
                (TFloat g_mode, TFloat g_mode) in
          let res_ty = ensure_float expected in
          let a_t = infer env a a_ty in
          let b_t = infer env b b_ty in
          { expr = EMul (a_t, b_t); typ = res_ty }
    | Div (a, b) ->
        let is_scaling = match (a,b) with (_, Const _) -> true | _ -> false in
        let a_ty, b_ty =
            if is_scaling then expected, expected
            else
            let g_mode = fresh_mode_meta () in
            set_mode g_mode G;
            (TFloat g_mode, TFloat g_mode)
        in
        let res_ty = ensure_float expected in
        let a_t = infer env a a_ty in
        let b_t = infer env b b_ty in
        { expr = EDiv (a_t, b_t); typ = res_ty }
  | Lt (a, b) ->
      let float_g = TFloat (fresh_mode_meta ()) in
      set_mode (match float_g with TFloat m -> m | _ -> assert false) G;
      let a_t = infer env a float_g in
      let b_t = infer env b float_g in
      assert_subtype TBool expected;
      { expr = ELt (a_t, b_t); typ = TBool }
  | Uniform (a, b) ->
      let ty = ensure_float expected in
      let a_t = infer env a ty in
      let b_t = infer env b ty in
      { expr = EUniform (a_t, b_t); typ = ty }
  | Gauss (a, b) ->
      let mean_ty = ensure_float expected in
      let var_mode = fresh_mode_meta () in
      set_mode var_mode G;
      let var_ty = TFloat var_mode in
      let a_t = infer env a mean_ty in
      let b_t = infer env b var_ty in
      { expr = EGauss (a_t, b_t); typ = mean_ty }
  | Exponential e ->
        let ty = ensure_float expected in
        let rate_mode = fresh_mode_meta () in
        set_mode rate_mode G;
        let rate_ty = TFloat rate_mode in
        let e_t = infer env e rate_ty in
        { expr = EExponential e_t; typ = ty }
  | Gamma (a, b) ->
        let a_ty = ensure_float expected in 
        let b_mode = fresh_mode_meta () in 
        set_mode b_mode G;
        let b_ty = TFloat b_mode in
        let a_t = infer env a a_ty in
        let b_t = infer env b b_ty in
        { expr = EGamma (a_t, b_t); typ = a_ty }
  | Beta (a, b) ->
        let ty = ensure_float expected in
        let param_mode = fresh_mode_meta () in
        set_mode param_mode G;
        let param_ty = TFloat param_mode in
        let a_t = infer env a param_ty in
        let b_t = infer env b param_ty in
        { expr = EBeta (a_t, b_t); typ = ty }
  | Flip p ->
      let p_ty = fresh_float () in
      let p_t = infer env p p_ty in
      assert_subtype TBool expected;
      { expr = EFlip p_t; typ = TBool }
  | Discrete choices ->
      let ty = ensure_float expected in 
      let typed_choices =
          List.map
            (fun (p, ei) ->
                if p < 0.0 || p > 1.0 then failwith "Discrete: probability not in [0,1]";
                let ei_t = infer env ei ty in
                (p, ei_t))
            choices
      in
      { expr = EDiscrete typed_choices; typ = ty }
      

File: sim/src/compiler/infer.js
import { CompileError } from "./errors.js";
import {
  TArrow,
  TBool,
  TFloat,
  TList,
  TMeta,
  TPair,
  TSum,
  TUnit,
  assertSubtype,
  defaultModesType,
  ensureFloat,
  formatType,
  freshFloat,
  freshMeta,
  freshModeMeta,
  resetTypeState,
  setMode,
  zonk,
} from "./types.js";

function typed(expr, typ, extra = {}) {
  return { kind: expr.kind, typ, from: expr.from, to: expr.to, ...extra };
}

function lookup(env, name, source) {
  if (!env.has(name)) throw new CompileError(`unbound variable \`${name}\``, source.from, source.to);
  return env.get(name);
}

function extend(env, entries) {
  const next = new Map(env);
  for (const [name, typ] of entries) next.set(name, typ);
  return next;
}

function forceAnnotatedMode(expr, typ) {
  if (!expr.mode) return;
  const floatTy = ensureFloat(typ, expr);
  setMode(floatTy.mode, expr.mode, expr);
}

function floatG() {
  const mode = freshModeMeta();
  setMode(mode, "G");
  return TFloat(mode);
}

export function inferProgram(expr) {
  resetTypeState();
  return infer(new Map(), expr, TMeta(freshMeta()));
}

export function infer(env, expr, expected) {
  switch (expr.kind) {
    case "Var": {
      const tyVar = lookup(env, expr.name, expr);
      assertSubtype(tyVar, expected, expr);
      return typed(expr, expected, { name: expr.name });
    }
    case "Lam": {
      const dom = TMeta(freshMeta());
      const cod = TMeta(freshMeta());
      const body = infer(extend(env, [[expr.param, dom]]), expr.body, cod);
      const lamTy = TArrow(dom, cod);
      assertSubtype(lamTy, expected, expr);
      return typed(expr, lamTy, { param: expr.param, body });
    }
    case "Rec": {
      const dom = TMeta(freshMeta());
      const cod = TMeta(freshMeta());
      const fnTy = TArrow(dom, cod);
      const body = infer(extend(env, [[expr.name, fnTy], [expr.param, dom]]), expr.body, cod);
      assertSubtype(body.typ, cod, expr.body);
      assertSubtype(fnTy, expected, expr);
      return typed(expr, fnTy, { name: expr.name, param: expr.param, body });
    }
    case "App": {
      const argTy = TMeta(freshMeta());
      const resTy = TMeta(freshMeta());
      const fnTy = TArrow(argTy, resTy);
      const fn = infer(env, expr.fn, fnTy);
      const arg = infer(env, expr.arg, argTy);
      assertSubtype(resTy, expected, expr);
      return typed(expr, resTy, { fn, arg });
    }
    case "Unit":
      assertSubtype(TUnit, expected, expr);
      return typed(expr, TUnit);
    case "Nil": {
      const elem = TMeta(freshMeta());
      const listTy = TList(elem);
      assertSubtype(listTy, expected, expr);
      return typed(expr, listTy);
    }
    case "Cons": {
      const elem = TMeta(freshMeta());
      const listTy = TList(elem);
      const head = infer(env, expr.head, elem);
      const tail = infer(env, expr.tail, listTy);
      assertSubtype(listTy, expected, expr);
      return typed(expr, listTy, { head, tail });
    }
    case "Pair": {
      const leftTy = TMeta(freshMeta());
      const rightTy = TMeta(freshMeta());
      const left = infer(env, expr.left, leftTy);
      const right = infer(env, expr.right, rightTy);
      const pairTy = TPair(left.typ, right.typ);
      assertSubtype(pairTy, expected, expr);
      return typed(expr, pairTy, { left, right });
    }
    case "Fst": {
      const a = TMeta(freshMeta());
      const b = TMeta(freshMeta());
      const exprTyped = infer(env, expr.expr, TPair(a, b));
      assertSubtype(a, expected, expr);
      return typed(expr, a, { expr: exprTyped });
    }
    case "Snd": {
      const a = TMeta(freshMeta());
      const b = TMeta(freshMeta());
      const exprTyped = infer(env, expr.expr, TPair(a, b));
      assertSubtype(b, expected, expr);
      return typed(expr, b, { expr: exprTyped });
    }
    case "Inl": {
      const leftTy = TMeta(freshMeta());
      const rightTy = TMeta(freshMeta());
      const value = infer(env, expr.expr, leftTy);
      const sumTy = TSum(value.typ, rightTy);
      assertSubtype(sumTy, expected, expr);
      return typed(expr, sumTy, { expr: value });
    }
    case "Inr": {
      const leftTy = TMeta(freshMeta());
      const rightTy = TMeta(freshMeta());
      const value = infer(env, expr.expr, rightTy);
      const sumTy = TSum(leftTy, value.typ);
      assertSubtype(sumTy, expected, expr);
      return typed(expr, sumTy, { expr: value });
    }
    case "Case": {
      const leftTy = TMeta(freshMeta());
      const rightTy = TMeta(freshMeta());
      const scrutinee = infer(env, expr.scrutinee, TSum(leftTy, rightTy));
      const left = infer(extend(env, [[expr.leftName, leftTy]]), expr.left, expected);
      const right = infer(extend(env, [[expr.rightName, rightTy]]), expr.right, expected);
      return typed(expr, expected, {
        scrutinee,
        leftName: expr.leftName,
        left,
        rightName: expr.rightName,
        right,
      });
    }
    case "MatchList": {
      const elemTy = TMeta(freshMeta());
      const listTy = TList(elemTy);
      const scrutinee = infer(env, expr.scrutinee, listTy);
      const nilBranch = infer(env, expr.nilBranch, expected);
      const consBranch = infer(extend(env, [[expr.headName, elemTy], [expr.tailName, listTy]]), expr.consBranch, expected);
      return typed(expr, expected, {
        scrutinee,
        nilBranch,
        headName: expr.headName,
        tailName: expr.tailName,
        consBranch,
      });
    }
    case "Bool":
      assertSubtype(TBool, expected, expr);
      return typed(expr, TBool, { value: expr.value });
    case "If": {
      const cond = infer(env, expr.cond, TBool);
      const thenBranch = infer(env, expr.thenBranch, expected);
      const elseBranch = infer(env, expr.elseBranch, expected);
      return typed(expr, expected, { cond, thenBranch, elseBranch });
    }
    case "Let": {
      const valueTy = TMeta(freshMeta());
      const value = infer(env, expr.value, valueTy);
      const body = infer(extend(env, [[expr.name, value.typ]]), expr.body, expected);
      return typed(expr, body.typ, { name: expr.name, value, body });
    }
    case "Const": {
      const ty = freshFloat();
      assertSubtype(ty, expected, expr);
      return typed(expr, ty, { value: expr.value });
    }
    case "Neg": {
      const ty = ensureFloat(expected, expr);
      const value = infer(env, expr.expr, ty);
      return typed(expr, ty, { expr: value });
    }
    case "Add":
    case "Sub": {
      const ty = ensureFloat(expected, expr);
      const left = infer(env, expr.left, ty);
      const right = infer(env, expr.right, ty);
      return typed(expr, ty, { left, right });
    }
    case "Mul": {
      const scaling = expr.left.kind === "Const" || expr.right.kind === "Const";
      if (scaling) {
        const ty = ensureFloat(expected, expr);
        return typed(expr, ty, { left: infer(env, expr.left, ty), right: infer(env, expr.right, ty) });
      }
      const gTy = floatG();
      const left = infer(env, expr.left, gTy);
      const right = infer(env, expr.right, gTy);
      const resTy = ensureFloat(expected, expr);
      return typed(expr, resTy, { left, right });
    }
    case "Div": {
      const scaling = expr.right.kind === "Const";
      const aTy = scaling ? expected : floatG();
      const bTy = scaling ? expected : aTy;
      const resTy = ensureFloat(expected, expr);
      const left = infer(env, expr.left, aTy);
      const right = infer(env, expr.right, bTy);
      return typed(expr, resTy, { left, right });
    }
    case "Lt":
    case "Leq": {
      const gTy = floatG();
      const left = infer(env, expr.left, gTy);
      const right = infer(env, expr.right, gTy);
      assertSubtype(TBool, expected, expr);
      return typed(expr, TBool, { left, right });
    }
    case "Uniform": {
      const ty = ensureFloat(expected, expr);
      forceAnnotatedMode(expr, ty);
      return typed(expr, ty, { mode: expr.mode, args: [infer(env, expr.args[0], ty), infer(env, expr.args[1], ty)] });
    }
    case "Gauss": {
      const meanTy = ensureFloat(expected, expr);
      forceAnnotatedMode(expr, meanTy);
      const args = [infer(env, expr.args[0], meanTy), infer(env, expr.args[1], floatG())];
      return typed(expr, meanTy, { mode: expr.mode, args });
    }
    case "Exponential": {
      const ty = ensureFloat(expected, expr);
      forceAnnotatedMode(expr, ty);
      return typed(expr, ty, { mode: expr.mode, args: [infer(env, expr.args[0], floatG())] });
    }
    case "Gamma": {
      const ty = ensureFloat(expected, expr);
      forceAnnotatedMode(expr, ty);
      return typed(expr, ty, { mode: expr.mode, args: [infer(env, expr.args[0], ty), infer(env, expr.args[1], floatG())] });
    }
    case "Beta": {
      const ty = ensureFloat(expected, expr);
      forceAnnotatedMode(expr, ty);
      const paramTy = floatG();
      return typed(expr, ty, { mode: expr.mode, args: [infer(env, expr.args[0], paramTy), infer(env, expr.args[1], paramTy)] });
    }
    case "Flip": {
      const p = infer(env, expr.args[0], freshFloat());
      assertSubtype(TBool, expected, expr);
      return typed(expr, TBool, { mode: expr.mode, args: [p] });
    }
    case "Bernoulli":
    case "Poisson": {
      const ty = ensureFloat(expected, expr);
      forceAnnotatedMode(expr, ty);
      return typed(expr, ty, { mode: expr.mode, args: [infer(env, expr.args[0], ty)] });
    }
    case "Discrete": {
      const ty = ensureFloat(expected, expr);
      forceAnnotatedMode(expr, ty);
      const choices = expr.choices.map((choice) => {
        if (choice.probability < 0 || choice.probability > 1) {
          throw new CompileError("discrete probability must be in [0, 1]", expr.from, expr.to);
        }
        return { probability: choice.probability, value: infer(env, choice.value, ty) };
      });
      return typed(expr, ty, { mode: expr.mode, choices });
    }
    case "Observe": {
      const cond = infer(env, expr.cond, TBool);
      assertSubtype(TUnit, expected, expr);
      return typed(expr, TUnit, { cond });
    }
    default:
      throw new CompileError(`unsupported expression kind ${expr.kind}`, expr.from, expr.to);
  }
}

export function defaultModes(typedExpr) {
  const go = (te) => {
    defaultModesType(te.typ);
    for (const child of typedChildren(te)) go(child);
  };
  go(typedExpr);
  return typedExpr;
}

export function typedChildren(te) {
  switch (te.kind) {
    case "Lam":
    case "Rec":
      return [te.body];
    case "App":
      return [te.fn, te.arg];
    case "Pair":
    case "Add":
    case "Sub":
    case "Mul":
    case "Div":
    case "Lt":
    case "Leq":
      return [te.left, te.right];
    case "Fst":
    case "Snd":
    case "Inl":
    case "Inr":
    case "Neg":
      return [te.expr];
    case "Cons":
      return [te.head, te.tail];
    case "Case":
      return [te.scrutinee, te.left, te.right];
    case "MatchList":
      return [te.scrutinee, te.nilBranch, te.consBranch];
    case "If":
      return [te.cond, te.thenBranch, te.elseBranch];
    case "Let":
      return [te.value, te.body];
    case "Uniform":
    case "Gauss":
    case "Exponential":
    case "Gamma":
    case "Beta":
    case "Flip":
    case "Bernoulli":
    case "Poisson":
      return te.args;
    case "Discrete":
      return te.choices.map((choice) => choice.value);
    case "Observe":
      return [te.cond];
    default:
      return [];
  }
}

export function collectSpans(te, spans = []) {
  spans.push({
    from: te.from,
    to: te.to,
    kind: ["Var"].includes(te.kind) ? "identifier" : isDistribution(te.kind) ? "distribution" : "expr",
    type: formatType(te.typ),
    mode: zonk(te.typ)?.tag === "Float" ? (zonk(te.typ).mode.mode ?? "?") : undefined,
    text: hoverText(te),
  });
  for (const child of typedChildren(te)) collectSpans(child, spans);
  return spans;
}

function hoverText(te) {
  const base = `${te.kind}: ${formatType(te.typ)}`;
  if (!isDistribution(te.kind)) return base;
  const mode = te.typ.tag === "Float" ? te.typ.mode.mode ?? "?" : "?";
  if (mode === "E") return `${base}\ndeterminizes to its expectation`;
  if (mode === "G") return `${base}\nsampled normally`;
  return base;
}

function isDistribution(kind) {
  return ["Uniform", "Gauss", "Exponential", "Gamma", "Beta", "Flip", "Bernoulli", "Poisson", "Discrete"].includes(kind);
}
